File: go/buscaBinaria/buscaBinaria.go
package buscaBinaria

func buscaBinaria(array_de_numeros []int, numero_buscado int) (string, int) {
	l := 0
	r := len(array_de_numeros) - 1

	for l <= r {
		meio := (l + r) / 2

		if array_de_numeros[meio] == numero_buscado {
			return "Busca Binária: número %v encontrado :)\n", numero_buscado
		}

		if array_de_numeros[meio] < numero_buscado {
			l = meio + 1
		} else {
			r = meio - 1
		}
	}

	return "Busca Binária: número %v não encontrado :(\n", numero_buscado
}

func RespostaBuscaBinaria() (string, int) {
	array_de_numeros := []int{1, 2, 3, 4, 5}
	numero_buscado := 5
	return buscaBinaria(array_de_numeros, numero_buscado)
}

/*
 * REFERÊNCIAS
 *		[Exemplo de Busca Binária](https://blog.geekhunter.com.br/golang/#:~:text=Busca binária com testes unitários para Golang,-O algoritmo de&text=Ao declarar variáveis com um,o valor do lado direito.)
 */

File: rust/src/busca_binaria.rs
fn retorna_busca_binaria<'a>(
    array_numeros: &'a [i32],
    numero_buscado: &'a i32,
) -> (&'static str, &'a i32) {
    let mut inicio = 0;
    let mut fim = array_numeros.len() - 1;

    while inicio <= fim {
        let meio = (inicio + fim) / 2;

        if array_numeros[meio] == *numero_buscado {
            return ("Número encontrado :)", &numero_buscado);
        }

        if array_numeros[meio] < *numero_buscado {
            inicio = meio + 1
        } else {
            fim = meio - 1
        }
    }
    ("Número não encontrado :(", &numero_buscado)
}

pub fn busca_binaria() -> () {
    let array_numeros = [1, 2, 3, 4, 5];
    let numero_buscado = 5;
    let (resposta, _) = retorna_busca_binaria(&array_numeros, &numero_buscado);
    println!("{} {}", resposta, &numero_buscado);
    ()
}
